Pull upstream

zuzunker opened the pull request:
Pulling latest upstream after https://github.com/input-output-hk/rust-cardano/pull/704

edolstra committed:
Fix above_stake_threshold

Since phi() is the probability that a pool is selected, we have to
check whether we're *below* that number.

edolstra committed:
Add some code for testing genesis leadership selection

edolstra committed:
Mark the phi test as "ignore" and add some assertions

zuzunker committed:
Extracted functions to work with lists of offsets written into a file, just for future code reusability

edolstra committed:
Don't make Slot.0 public

vincenthz committed:
Merge pull request #697 from input-output-hk/phi-tests

Fix above_stake_threshold and add genesis leadership selection test

vincenthz committed:
Merge pull request #704 from Emurgo/extracted-file-offset-functions

Extracted file-offset functions

zuzunker committed:
Merge remote-tracking branch 'remotes/origin/master-emurgo' into pull-upstream

rooooooooob reviewed (approved):
lgtm, just mockchain changes.

nicarq commented on chain-impl-mockchain/src/leadership/genesis/mod.rs:
```diff
@@ -136,6 +136,137 @@ impl GenesisLeaderSelection {
+            (big_pool_vrf_private_key, 0, Value(1000)),
+        );
+
+        for _i in 0..10 {
```
you can just do `_` in Rust :) 

https://play.rust-lang.org/?code=fn%20main()%0A%7B%0A%20%20%20%20for%20_%20in%200..10%20%7B%0A%20%20%20%20%20%20%20%20println!(%22Hello%20World%22)%3B%0A%20%20%20%20%7D%0A%7D&version=stable

zuzunker replied:
That's not our code, tho )

nicarq commented on chain-impl-mockchain/src/leadership/genesis/mod.rs:
```diff
@@ -136,6 +136,137 @@ impl GenesisLeaderSelection {
+
+        let nr_slots = slots_per_epoch;
+
+        for _i in 0..nr_slots {
```
same as above

nicarq commented on chain-impl-mockchain/src/leadership/genesis/mod.rs:
```diff
@@ -136,6 +136,137 @@ impl GenesisLeaderSelection {
+        let slots_per_epoch = 200000;
+        let active_slots_coeff = 0.1;
+
+        let (_genesis_hash, mut ledger) = create_initial_fake_ledger(
```
if this is not used, I think you can also just do `_`

rooooooooob replied:
It's not used. marking it as _genesis_hash is done to both signify it is not used (so it doesn't trip the rustc unused variable linter), while also still documenting what it is we're iterating over.

This PR is also just updating our fork with IOHK's repo (thus this is their code) before we send them a PR back, so we don't want to be making needless changes which would just complicate our PRs in their repo.

All changes in this PR are also in the in-development/experimental shelley module which we are not using yet for the bridge or anything else.

Commit:
Add example of transaction signing

Add function to get txid from not yet signed transaction (because it's
needed for signing)
Add function to get root xprv (to be able to derive the necessary
private key for signing)

Commit:
Documentation

Commit:
add missing memory deallocation

Commit:
replace deprecated method

vincenthz committed:
remove fakemmm

NicolasDP committed:
split the apply_block and apply_message

while this function should not be used to create a new ledger state
it is necessary to have for the block content selection

NicolasDP committed:
add LedgerParameters in the Leadership object

vincenthz committed:
Add KeyEvolvingAlgorithm

vincenthz committed:
Update interfaces for KeyEvolvingAlgorithm

vincenthz committed:
adjust interface for block creation to use normal signing

vincenthz committed:
rustfmt

vincenthz committed:
add leadereship date_at_slot

vincenthz committed:
Add documentation to leadership parameters

vincenthz committed:
Merge pull request #707 from input-output-hk/mockchain-changes

Mockchain changes: LedgerParameters in the Leadership and apply_fragment

vincenthz committed:
add last comment

vincenthz committed:
move period to u32 for now

vincenthz committed:
Merge pull request #709 from input-output-hk/leadership-helper-doc

Leadership Helper and Doc

NicolasDP committed:
Merge pull request #708 from input-output-hk/key-evolving-cleanup

key evolution cleanup

vincenthz committed:
Merge pull request #706 from atixlabs/IOHKCRB-34-Sign-transaction-example

IOHKCRB-34 Sign transaction example

NicolasDP committed:
general improvements in the error reporting

NicolasDP committed:
remove debug prints

vincenthz committed:
adjust documentation of chain-addr

NicolasDP committed:
Merge pull request #710 from input-output-hk/errors

Errors improvements

vincenthz committed:
crypto: make public key surjective

multiple secret key maps their public key to the same public key algorithm.
Allow to specify public key as ed25519, regardless
of the type of ed25519 (normal or extended)

vincenthz committed:
update interface from chain-addr to use ed25519

vincenthz committed:
change mockchain for new crypto separation

use Ed25519 in public interfaces now
adjustment linked to the new split in half constraint to traits

sjmackenzie committed:
AddressReadable prints AddressReadable("<addr>")

This fix ensures the actual output is "<addr>".

vincenthz committed:
rustfmt

vincenthz committed:
Merge pull request #712 from sjmackenzie/address_readable_to_string

AddressReadable prints AddressReadable("<addr>")

NicolasDP committed:
Merge pull request #713 from input-output-hk/crypto-pubkey-surjective

Make pubkey surjective

vincenthz committed:
register stake key on auto creation of account

vincenthz committed:
Merge pull request #715 from input-output-hk/bug-account-autocreate

register stake key on auto creation of account

sjmackenzie committed:
remove unused imports

Commit:
Merge pull request #716 from sjmackenzie/remove-unused-imports

remove unused imports

Commit:
Bump tower-grpc to latest rev supporting tower-h2

After this revision, the "tower-h2" feature no longer exists.

Commit:
network-core: Sendable server response streams

Require that the response streams produced by the service
implementation satisfy Send + 'static. This reduces a lot of bound
work for generic protocol implementations that require the streams
to be sendable across threads. A future port of network-grpc to
tower-hyper will be one such implementation.

Commit:
networ-core: Remove method propose_transactions

This ContentService method was only used by network-ntt, and, if that
protocol implementation is ever continued, it should be replaced with
the implementation of message_subscription.

Commit:
network-core: sendable service futures

Require that any response futures the service implementation
returns satisfy Send + 'static, just like the streams.

Commit:
network-core: sendable P2pService::NodeId

The tower-hyper port needs this, too.

NicolasDP committed:
Merge pull request #718 from input-output-hk/network-service-returns-sendable

Send + 'static bounds for types returned by the network service implementation

zuzunker committed:
Merge remote-tracking branch 'remotes/origin/master' into pull-upstream

zuzunker committed:
Merge remote-tracking branch 'remotes/origin/master-emurgo' into pull-upstream
